### PR TITLE
Andrei/cancel query

### PR DESCRIPTION
**Pull request by AndreiBozantan**
Added functionality to cancel explorations and underlying Aircloak API queries .

### CODE REVIEW

File: src/explorer.api/Controllers/ExploreController.cs [hunk 1]
```diff
@@ -126,33 +120,47 @@ public async Task<IActionResult> Result(System.Guid explorationId)
...
 
             var components = type switch
             {
                 AircloakType.Integer => new ExplorerBase[]
                 {
-                    new IntegerColumnExplorer(resolver, data.TableName, data.ColumnName),
-                    new MinMaxExplorer(resolver, data.TableName, data.ColumnName),
+                    new IntegerColumnExplorer(resolver, data.TableName, data.ColumnName, cts.Token),
+                    new MinMaxExplorer(resolver, data.TableName, data.ColumnName, cts.Token),
                 },
```
Comment (dandanlen) [inline src/explorer.api/Controllers/ExploreController.cs]: Ok.. I can see how creating the cts.Token in the Exploration might be tricky. 

We may have to have to move this switch statement into the `Exploration` constructor, otherwise we can pass delegates instead of Explorer instances. Something like: 
```csharp
var components = (token) => new ExplorerBase[]
                {
                    new IntegerColumnExplorer(resolver, data.TableName, data.ColumnName, token),
                    new MinMaxExplorer(resolver, data.TableName, data.ColumnName, token),
                },

[etc.]
```

and in the Exlporation constructor:
```csharp
Exploration(Func<Token, ExplorerBase[]> explorerBuilder)
{
    var tokenSource = new CancellationTokenSource();
    var components = explorerBuilder(tokenSource.Token);
[etc.]
}
```

File: tests/explorer.api.tests/TestWebAppFactory.cs [hunk 2]
```diff
@@ -101,9 +101,11 @@ public FileInfo GetVcrCasetteInfo(string testClassName, string vcrSessionName)
             return new FileInfo($"../../../.vcr/{testClassName}.{vcrSessionName}.yaml");
         }
 
-        public TimeSpan? GetApiPollingFrequency(FileInfo vcrCassetteInfo)
+        public TimeSpan GetApiPollingFrequency(FileInfo vcrCassetteInfo)
         {
-            return (vcrCassetteInfo.Exists && vcrCassetteInfo.Length > 0) ? TimeSpan.FromMilliseconds(1) : default(TimeSpan?);
+            return (vcrCassetteInfo.Exists && vcrCassetteInfo.Length > 0) ?
```
Comment (dandanlen) [inline tests/explorer.api.tests/TestWebAppFactory.cs]: Should maybe check that we are no in VcrMode here - if the file exists and vcr is in record mode, we could end up hammering the api every millisecond...

File: src/aircloak/JsonApi/JsonApiClient.cs [hunk 3]
```diff
@@ -320,37 +288,41 @@ private static string ServiceError(HttpResponseMessage response)
...
                     contentStream,
-                    options ?? DefaultJsonOptions);
+                    options,
+                    cancellationToken);
             }
             else
             {
-                var responseContent = await response.Content.ReadAsStringAsync();
+                var responseContent = await Task.Run(response.Content.ReadAsStringAsync, cancellationToken);
+                responseContent = await Task.Run(response.Content.ReadAsStringAsync, cancellationToken);
```
[the remark this answers is not in the document]
Comment (AndreiBozantan) [reply src/aircloak/JsonApi/JsonApiClient.cs]: fixed with d2c4bd91e933c8ee09237efcdf489a91e052c3e3

File: src/aircloak/JsonApi/JsonApiClient.cs [hunk 4]
```diff
@@ -104,137 +97,108 @@ public async Task<DataSourceCollection> GetDataSources()
...
             {
-                if (e.CancellationToken == tokenSource.Token)
+                if (ex.CancellationToken == cancellationToken)
                 {
-                    throw new TimeoutException($"Timed out while waiting for query results for {queryId}");
+                    await CancelQuery(queryId, cancellationToken);
                 }
-
                 throw;
             }
```
[the remark this answers is not in the document]
Comment (AndreiBozantan) [reply src/aircloak/JsonApi/JsonApiClient.cs]: I actually modified the code to cancel the Aircloak API query in case of any exception. 
I am not sure the IsCancellationRequested is actually helpful, because the cancellation request can be also detected by the methods called from PollQueryUntilComplete, so we will just get an OperationCancelledException, without the chance of checking the IsCancellationRequested outside of a catch block.
da6c5c0bc330a616c508dbd67cedf68898a1d2ec

File: src/aircloak/JsonApi/JsonApiClient.cs [hunk 5]
```diff
@@ -104,137 +97,108 @@ public async Task<DataSourceCollection> GetDataSources()
...
+                    {
+                        switch (queryResult.Query.QueryState)
+                        {
+                            case "completed":
+                                return queryResult;
+                            case "error":
+                                throw new Exception("Aircloak API query error.\n" +
+                                    GetQueryResultDetails(querySpec, queryResult));
+                            case "cancelled":
+                                throw new OperationCanceledException("Aircolak API query canceled.\n" +
```
[the remark this answers is not in the document]
Comment (AndreiBozantan) [reply src/aircloak/JsonApi/JsonApiClient.cs]: fixed ffcdea79845d9f362c7ddc307b8b614190e79cfa

File: src/aircloak/JsonApi/JsonApiClient.cs [hunk 6]
```diff
@@ -104,137 +97,108 @@ public async Task<DataSourceCollection> GetDataSources()
...
+                                    GetQueryResultDetails(querySpec, queryResult));
+                        }
+                    }
 
-            try
-            {
-                return await PollQueryUntilComplete<TRow>(queryId, querySpec, tokenSource.Token, pollFrequency);
+                    cancellationToken.ThrowIfCancellationRequested();
+                    await Task.Delay(pollFrequency, cancellationToken);
+                    cancellationToken.ThrowIfCancellationRequested();
```
[the remark this answers is not in the document]
Comment (AndreiBozantan) [reply src/aircloak/JsonApi/JsonApiClient.cs]: fixed 16e0d7aa209f50fdb90436bdf667fd663fbcfb3b

File: src/aircloak/JsonApi/JsonApiClient.cs [hunk 7]
```diff
@@ -104,137 +97,108 @@ public async Task<DataSourceCollection> GetDataSources()
...
         /// <summary>
         /// Sends a Http GET request to the /api/queries/{query_id}/cancel. Cancels a running query on the Aircloak
         /// server.
         /// </summary>
         /// <param name="queryId">The id of the query to cancel.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken" /> object that can be used to cancel the operation.</param>
         /// <returns>A <c>CancelResponse</c> instance indicating whether or not the query was indeed canceled.
         /// </returns>
-        public async Task<CancelResponse> CancelQuery(string queryId)
+        public async Task<CancelResponse> CancelQuery(string queryId, CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (AndreiBozantan) [reply src/aircloak/JsonApi/JsonApiClient.cs]: fixed with da6c5c0bc330a616c508dbd67cedf68898a1d2ec

File: src/explorer.api/Controllers/ExploreController.cs [hunk 8]
```diff
@@ -41,7 +45,8 @@ public async Task<IActionResult> Explore(Models.ExploreParams data)
         {
             authProvider.RegisterApiKey(data.ApiKey);
 
-            var dataSources = await apiClient.GetDataSources();
+            using var cts = new CancellationTokenSource();
+            var dataSources = await apiClient.GetDataSources(cts.Token);
```
[the remark this answers is not in the document]
Comment (AndreiBozantan) [reply src/explorer.api/Controllers/ExploreController.cs]: fixed with 60fa1549b948208e2d7edcecb13fbf6bd2a83179

File: src/explorer.api/Explorers/Exploration.cs [hunk 9]
```diff
@@ -3,17 +3,20 @@ namespace Explorer
...
 
-        public Exploration(IEnumerable<ExplorerBase> explorers)
-        {
-            ExplorationGuid = Guid.NewGuid();
+        private readonly CancellationTokenSource cancellationTokenSource;
 
+        public Exploration(IEnumerable<ExplorerBase> explorers, CancellationTokenSource cancellationTokenSource)
+        {
             this.explorers = explorers;
+            this.cancellationTokenSource = cancellationTokenSource;
```
[the remark this answers is not in the document]
Comment (AndreiBozantan) [reply src/explorer.api/Explorers/Exploration.cs]: fixed with 60fa1549b948208e2d7edcecb13fbf6bd2a83179

File: src/aircloak/JsonApi/JsonApiClient.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dandanlen) [reply src/aircloak/JsonApi/JsonApiClient.cs]: Ok, that makes sense, I assumed that a canceled Task would simply return normally in a canceled state, didn't realise it threw an exception.

File: src/aircloak/JsonApi/JsonApiClient.cs [hunk 10]
```diff
@@ -135,15 +135,23 @@ public async Task<DataSourceCollection> GetDataSources(CancellationToken cancell
...
+            var speed = 16;
+
             try
             {
                 while (true)
                 {
+                    await Task.Delay(pollFrequency / speed, cancellationToken);
+
+                    cancellationToken.ThrowIfCancellationRequested();
+
```
Comment (dandanlen) [inline src/aircloak/JsonApi/JsonApiClient.cs]: Why did you move the delay to the beginning of the loop? Surely this means there's a (short) delay before even sending the request for the first time?

File: src/explorer.api/Controllers/ExploreController.cs (hunk 1 above)
Comment (AndreiBozantan) [reply src/explorer.api/Controllers/ExploreController.cs]: fixed with 60fa1549b948208e2d7edcecb13fbf6bd2a83179

File: tests/explorer.api.tests/QueryTests.cs [hunk 11]
```diff
@@ -169,6 +170,37 @@ public async void TestRepeatingRows()
             });
         }
 
+        [Fact]
+        public async void TestCancelQuery()
+        {
+            using var client = new HttpClient() { BaseAddress = TestWebAppFactory.Config.AircloakApiUrl };
```
Comment (dandanlen) [inline tests/explorer.api.tests/QueryTests.cs]: You could use the VCR here. 

```suggestion
            var vcrCassetteInfo = factory.GetVcrCasetteInfo(nameof(QueryTests), nameof(TestCancelQuery));
            using var client = factory.CreateAircloakApiHttpClient(vcrCassetteInfo);
```

File: tests/explorer.api.tests/QueryTests.cs [hunk 12]
```diff
@@ -169,6 +170,37 @@ public async void TestRepeatingRows()
...
+            var authProvider = factory.EnvironmentVariableAuthProvider();
+            var pollFrequency = TimeSpan.FromMilliseconds(10);
+            var jsonApiClient = new JsonApiClient(client, authProvider);
+            var query = new Min("loans", "amount", null);
+
+            using var cts = new CancellationTokenSource();
+            var queryInfo = await jsonApiClient.SubmitQuery("gda_banking", query.QueryStatement, cts.Token);
+            cts.CancelAfter(1);
```
Comment (dandanlen) [inline tests/explorer.api.tests/QueryTests.cs]: This can simply be `cts.Cancel()`

File: tests/explorer.api.tests/TestWebAppFactory.cs (hunk 2 above)
Comment (AndreiBozantan) [reply tests/explorer.api.tests/TestWebAppFactory.cs]: Created issue #66 for this.

File: src/aircloak/JsonApi/JsonApiClient.cs (hunk 10 above)
Comment (AndreiBozantan) [reply src/aircloak/JsonApi/JsonApiClient.cs]: fixed with 6f0f7afeb172bae2cdbc80959addbd0b118eb40b

File: tests/explorer.api.tests/QueryTests.cs (hunk 11 above)
Comment (AndreiBozantan) [reply tests/explorer.api.tests/QueryTests.cs]: fixed with 1291bc2d3824a5c765573950679d8817613a7754

File: tests/explorer.api.tests/QueryTests.cs (hunk 12 above)
Comment (AndreiBozantan) [reply tests/explorer.api.tests/QueryTests.cs]: fixed with 1291bc2d3824a5c765573950679d8817613a7754
